### Copyright-protected material, all rights reserved. (c) University of Vienna.
_Copyright Notice of the corresponding course at Moodle applies. <br> Only to be used in the MRE course._

# MRE Assignment 3 - Digital Video Processing

In this assignment, you will use OpenCV and FFmpeg to implement very basic video editing functions. These tasks include:

1. Create a slide show (as a video) from images, and optionally create the slideshow as greyscale video.
2. Extract the audio track from a video file.
3. Replace the audio track in a video file.
4. Combine two or more videos into one video file.
5. Blend an image (fade-in/fade-out) with a video.
6. Blend two videos into one video (video collage).

Following, you find the detailed specification of the assignment. For the assessment of your solution you are expected to demonstrate your implementation by playing back the produced video files and - for the purpose of profound understanding - to explain the output FFmpeg generates.

❗ **Note:** To playback the produced videos in this assignment, please make use of the video-element of HTML5 to control the playback of a video so that the playback controls are available to the user.

    **Example:** 
    Input: `./resources/video/sample1.mp4`.
    Output: Video is played back 
    # the following function is to be provided by you
    Function call: `VideoPlayer("./resources/video/sample1.mp4")`

❗ **Note:** General requirements for output Videos where applicable as follows:
-	Output format: MP4
-	Output Aspect Ratio: 1280*720

❗ **Note:** Please make sure that all potential errors, including handling files, paths, and run-time errors are handled properly (e.g., useful error messages to users).

❗ **Note:** You must not use MoviePy in this assignment, as it causes some memory leaks and creates a lot of problems for you.

## Import your implementation

Import the corresponding Jupyter Notebook named "*_impl.ipynb" for this assignment here.

In [3]:
%%capture
%run MRE_A3_impl.ipynb

ffmpeg version 5.1.2-tessus Copyright (c) 2000-2022 the FFmpeg developers
  built with Apple clang version 11.0.0 (clang-1100.0.33.17)
  configuration: --cc=/usr/bin/clang --prefix=/opt/ffmpeg --extra-version=tessus --enable-avisynth --enable-fontconfig --enable-gpl --enable-libaom --enable-libass --enable-libbluray --enable-libdav1d --enable-libfreetype --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libmysofa --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopus --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvmaf --enable-libvo-amrwbenc --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxavs --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-version3 --pkg-config-flags=--static --disable-ffplay
  libavutil      57. 28.100

SyntaxError: invalid syntax (2943456693.py, line 4)

SyntaxError: invalid syntax (2943456693.py, line 4)

frame= 1656 fps=220 q=-1.0 Lsize=    5379kB time=00:00:58.49 bitrate= 753.3kbits/s speed=7.76x    
video:4401kB audio:910kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 1.284982%
[libx264 @ 0x7f837ef18500] frame I:21    Avg QP:19.82  size: 16550
[libx264 @ 0x7f837ef18500] frame P:435   Avg QP:22.64  size:  7264
[libx264 @ 0x7f837ef18500] frame B:1200  Avg QP:22.40  size:   832
[libx264 @ 0x7f837ef18500] consecutive B-frames:  1.7%  3.6%  4.3% 90.3%
[libx264 @ 0x7f837ef18500] mb I  I16..4: 12.4% 76.0% 11.6%
[libx264 @ 0x7f837ef18500] mb P  I16..4:  3.6% 16.8%  0.9%  P16..4: 41.1% 12.1%  7.3%  0.0%  0.0%    skip:18.1%
[libx264 @ 0x7f837ef18500] mb B  I16..4:  0.1%  0.4%  0.0%  B16..8: 16.8%  1.8%  0.4%  direct: 2.8%  skip:77.7%  L0:49.7% L1:43.5% BI: 6.8%
[libx264 @ 0x7f837ef18500] 8x8 transform intra:77.7% inter:77.4%
[libx264 @ 0x7f837ef18500] coded y,uvDC,uvAC intra: 54.6% 82.2% 20.7% inter: 9.0% 17.8% 0.3%
[libx264 @ 0x7f837ef18500] i16 v,h,dc,p: 41% 22%  4% 33%

## Task 3.1: Create a slide show (Video) from multiple images and convert it to greyscale - 15P

Write a Python function `CreateVideoFromImages` so that one can call it with the following parameters:

- `inImgLibFolder`: the input folder that contains the images required for the slide show. All images to be used for the slide show need to be of the same image format, e.g., png, or jpg.<br>
- `imageFormat`: image format available in the input folder.<br>
- `durationInSec`: duration of each image in seconds (all images are to be presented for the same duration).<br>
- `convertToGreyScale`: a boolean flag, set to `true` for producing a greyscale video.<br>
- `outFolder`: the output folder.<br>
- `outVideo`: the name of the output file.<br>


### Demonstrate your implementation:

In [ ]:
CreateVideoFromImages("task1/images/", "JPG", "2", True, "task1/output/", "outVideo")

## Task 3.2: Extract the Audio Track from a Video File - 15P

Given a video we would like to have the audio track of the video as a seperate audio file.
Implement the appropriate function:

Function `SplitAudioVideoTracks` with the following parameters:

-`inVideo`: the name of the input file (Video file).<br>
-`outFolder`: the output folder.<br>
-`outVideoTrack`: the name of the Video output file not containing any audio track anymore.<br>
-`outAudioTrack`: the name of the Audio output file representing the audio track of the input video.<br>


### Demonstrate your implementation:

In [ ]:
SplitAudioVideoTracks("./resources/video/sample1.mp4", "./task2/output/", "outVideoTrack.mp4", "outAudioTrack.wav")

## Task 3.3: Replace the Audio Track in a Video File - 15P

Given a video we would like to have an audio track added to the video. If the video has already an audio track, that track should be replaced. If the video does not yet have an audio track, the input audio track becomes the new audio track. 
Implement the appropriate function:

Function `AddOrReplaceAudio` with the following parameters:

-`inVideo`: the name of the input Video file.<br>
-`inAudio`: the new Audio file to be added to the Video.<br>
-`outFolder`: the output folder.<br>
-`outVideo`: the name of the output Video file.<br>


### Demonstrate your implementation:

In [ ]:
AddOrReplaceAudio("./resources/video/sample1.mp4", "./resources/audio/Amazon.mp3", "./task3/output/", "outTask3Video.mp4")

## Task 3.4: Combine Videos - 35P

Write a Python function `CombineVideos` so that one can call it with the following parameters:

- `inVideoLibFolder`: the input folder that contains the videos to be combined.<br>
- `outFolder`: the output folder.<br>
- `outVideo`: the name of the output video file.<br>

For this task, before combining the video files, the function will check and display the following information about each input video. To add more structure to your code, consider implementing a separate function `VideoMetadataExtractor`  to extract the required metadata:

* The video track:<br>
        o	videoCodec and codecID
        o	videoFrameRate [frames per second]
        o	videoLength [seconds]
        o	videoHeight [pixel]
        o	videoWidth [pixel]
        o	apectRatio [n:m]
<br>       
* The audio track: 
        o	audioCodec and codecID
        o	audioChannels (number of channels)
        o	audioSampleRate [Hz]
        o	audioBitRate [kb/s]

❗ **Note:** The function `VideoMetadataExtractor` displays a single pandas DataFrame representing the data for both tracks that can be displayed. The DataFrame represents a table with the columns of the metadata elements mentioned above.

The result might look like this (just a sample): <br>
![SampleTable](./A3T3_sampleTable.png)


💡 **TIP:** For processing the audio/video track of the video file consider using ffprobe available via FFmpeg. <br>


💡 **TIP:** Consider using `subprocess` to call FFmpeg-commands from Python. `subprocess` allows you to capture and print the output of FFmpeg-commands, which is very helpful for debugging and understanding the functionality.




### Demonstrate your implementation:

In [4]:
display(VideoMetadataExtractor("./resources/video/"))

filename vCodec   vcodecID       vDur       vFPS  vHeight  vWidth  \
0  sample1.mp4   avc1  828601953  42.375667  29.993628    240.0   352.0   
1  sample2.mp4   avc1  828601953   7.107100  30.110734    480.0   640.0   
2  sample3.avi   H264  875967048   7.173833  23.976024    486.0   720.0   

  aCodec        acodecID aChannels aSamplerate aBitrate  
0    aac            mp4a         2       44100   127761  
1    aac            mp4a         2       44100   128289  
2    aac  [255][0][0][0]         2       44100   128000

In [ ]:
CombineVideos("./resources/video/", "./task4/output/", "outTask4.mp4")
VideoPlayer("./task4/output/outTask4.mp4")

## Task 3.5: Blend an Image in a Video - 10P

Write a Python function `AddFadingImage` so that one can call it with the following parameters. 

- `inVideo`: the name of the input video file.<br>
- `inImg`: the image to be added.<br>
- `time`: the time N (in seconds) the image should be visible with full opacity, after fading in and before fading out.<br>
- `outFolder`: the output folder.<br>
- `outVideo`: the name of the output video file.<br>

❗ **Note:** The period P of fading in and fading out the image can be chosen freely, but P > 0. E.g., P = 0,01 is a feasible value.



### Demonstrate your implementation:

## Task 3.6: Create a Video Collage - blend two videos into one - 10P

Write a Python function `VideoClipMixer` so that one can call it with the following parameters:

- `inVideo1`: the input video 1.
- `inVideo2`: the input video 2.
- `layout` :  row (videos are placed left/right) vs. column (videos placed top/bottom).
- `outFolder`: the output folder.
- `outVideo`: the name of the output Video file.

Please note, unlike task 3.4, videos in this tasks should be played back simultaneously

### Demonstrate your implementation: